### ISSUE TITLE
Record types of some sort

### ISSUE THREAD

**Issue by hikari-no-yume** [TazeTSchnitzel/Firth#18]
absolutely no clue how these will work

**Comment by hikari-no-yume** [comment]
One idea: plain objects, JS-style (albeit without prototypes). Sort-of [like Emily](https://bitbucket.org/runhello/emily/src/47ab9350928d138dba4afa052cbce89108654709/doc/tutorial.md?at=default), you'd construct them by assigning to some variables within a scope, then taking the bindings in that scope and using them as the key/value pairs of the object. Thus:

``` firth
/mycircle [
    /x 75 def.
    /y 50 def.
    /radius 5 def.
    /colour "purple" def.
] object. def.
```

That'd result in `mycircle` being an object with the keys `x`, `y`, `radius` and `colour`.

I think this could work.

**Comment by hikari-no-yume** [comment]
I like how it goes with `list`. `object` takes the scope produced by a function and transforms it into an object. `list` takes the stack produced by a function and transforms it into a list.

**Comment by mathroc** [comment]
Couldn't it work with just :

```
/mycircle [
    /x 75
    /y 50
    /radius 5
    /colour "purple"
] object. def.
```

`object` could take a function and treat the output as a succession of `/key value`. `object` would execute the function with a new stack and use the stack content to build the object. it's less verbose and more similar to the `list` function but maybe a bit less clear and prone to error (if stack height is not even or if there is not a symbol when expected)

second point: how would you access values in the object. I was thinking : `/x mycircle.` not sure how to list the object keys without adding either another function (eg: `mycircle keys.`) or reserving a key in the object (eg: `/keys mycircle.`)
nb: the same goes for getting list values and list length

**Comment by hikari-no-yume** [comment]
Just using a succession of key-value pairs could work too, sure.

Also not sure about lookup. If an object call be called like a function can, then `/colour mycircle.` works, but you'd still need functions for other cases.

**Comment by mathroc** [comment]
yep, an alternative may be to something like that:
- `/x /get mycircle.` (-> value)
- `/keys mycircle.` (-> list)
- `2 /get mylist.` (-> value)
- `/length mylist.` (-> integer)
- `/pop mylist` (-> newlist, value)

**Comment by raoulvdberge** [comment]
How would a value from an object be changed?

**Comment by hikari-no-yume** [comment]
That is a good question. It mostly comes down to whether objects would be immutable, and so you'd create a new object with something changed, or mutable, so you'd modify an existing object.

This makes no difference if you're working on the stack. But it's quite a crucial difference if you're modifying a value accessible somewhere else.

**Comment by raoulvdberge** [comment]
+1 for immutability.

**Comment by mathroc** [comment]
We can imagine something like this to create a copy with one value changed:

```
/x 70 /with mycircle.
```

we don't have to repeat all existing value for that and it could easily be extended to change multiple values:

```
[/x 70 /y 55] /with mycircle.
```

I'm not sure about the `/with` keyboard it's just that I look at [PSR7](https://github.com/php-fig/fig-standards/blob/master/proposed/http-message.md) and it has `with*(…)` function for that

**Comment by raoulvdberge** [comment]
+1 for `/with`, but normal mutability should also be supported.

`/x 70 mycircle.` should just change `x` to `70` in `mycircle`, and `/x 70 /with mycircle.` should create a new instance of `mycircle` and update it, and then add the new instance to the stack.

**Comment by mathroc** [comment]
well, it can be ambiguous here : `/x /x mycircle.` should that means :
- set the value of `/x` in `mycircle` to `/x` and return the new record
- or, push `/x` on the stack and then push the value of `/x` in `mycircle` on the stack ?

**Comment by raoulvdberge** [comment]
Hmmm, true. Then normal mutability shouldn't be supported.

**Comment by raoulvdberge** [comment]
How would function invocation work for functions IN functions?

```
/circle [
  /x 0
  /y 0
  /hello [
    'Hello World from circle' print.
  ]
] object. def.

/hello circle..
```

`/hello circle.` would get the `hello` function in `circle`, and the last `.` would invoke it?

**Comment by hikari-no-yume** [comment]
I'm not actually sure about how methods should work, or if we should even have them. They don't seem too elegant so far, and you'd end up having to do double-invocations in places (`..`) which is weird. I suppose there are namespacing issues, but for now I'm fine with just functions. Haskell does that and it seems to work out okay.

I think the simplest thing would be `get` and `set`:
- `get` retrieves a named property
- `set`  returns a new object with some property added

So, to add 10 to the `x` property on a circle:

``` firth
/x circle get.
10 add.
/x circle set.
```

Or something like that.

:)

**Comment by hikari-no-yume** [comment]
You could have utility functions to make that specific one simpler. Like `update`, say, which might take a function:

``` firth
/x circle [
    10 add.
] update.
```

That could just be done in userland, or it could be built-in.

**Comment by raoulvdberge** [comment]
I agree, so an error should be thrown when trying to declare a function in a function?

**Comment by hikari-no-yume** [comment]
No, it'd work like it does everywhere else. It's local to that scope.

**Comment by raoulvdberge** [comment]
Alright. I really like the `get` / `set` stuff, much more simpler than having `/with`.

**Comment by hikari-no-yume** [comment]
Oh uh, just to be clear, yes you could do methods. I'm just not sure about using them for object manipulation at the moment.

We don't do it for other types, it prevents you making your own `get`/`set`/whatever methods to do something specific to that object, and at least for now methods seem to be quite ugly. But we'll see.

**Comment by raoulvdberge** [comment]
This would introduce a new type, object, right?

**Comment by hikari-no-yume** [comment]
Sure.

**Comment by mathroc** [comment]
from a user point of vue it should not be _very_ different if it's done with methods `<args> /method myrecord.` or functions `<args> myrecord function.` the only issue I have with functions is that it will declare a lot of functions that would become similar to reserved keywords and that some functions would have to work on different type of objects (list & record for `get` & `set`)

you certainly can build most utility functions in user-space but I can think a few I'd like to have in stdlib when programming (eg, for list: `contains`, `reverse`, `sort`, `length`, `pop`, `filter`, `reduce`, `map`, etc.)

**Comment by hikari-no-yume** [comment]
Yeah, I have the same issue with them. Lots of global functions is a problem. We'll see.

**Comment by raoulvdberge** [comment]
What about default values? Will we even support it? Will we have a null type?

```
/circle [
  /radius
  /x
  /y 0
] object. def.
```

This is problematic, because the programmer would think `/radius` has no value, but in fact, its value would be `/x`.

**Comment by mathroc** [comment]
I don't think records are supposed to have default values, records are independent, they are not object constructors

**Comment by raoulvdberge** [comment]
@mathroc I like your idea of doing multiple sets in once, when doing this, should that function first be converted to a list or not? What shall we do?

`[/x 70 /y 55] list. mycircle set.` or `[/x 70 /y 55] mycircle set.`
